packages/timeline-state-resolver/src/conductor.ts

```diff
@@ -1062,6 +1075,128 @@ export class Conductor extends EventEmitter<ConductorEvents> {
+			this._deviceStates[deviceId].reverse().find((s) => s.time <= this.getCurrentTime()),
+			...this._deviceStates[deviceId]
+				.reverse()
+				.filter((s) => s.time < time && s.time > this.getCurrentTime())
+				.reverse(),
+			{
+				time,
+				state,
+				dependencies,
+				mappings,
+			},
+		])
+
+		const filledState: typeof state = JSON.parse(JSON.stringify(state))
+		const fillState = (content: Record<string, any>, instance: TimelineObjectInstance) => {
+			Object.entries(content).forEach(([index, val]) => {
+				if (typeof val === 'object') {
+					if ('_datastoreKey' in val) {
+						const datastoreVal = this._datastore[val._datastoreKey]
+
+						if (!datastoreVal) {
+							content[index] = val.default
+						} else if (instance.originalStart && instance.originalStart > datastoreVal.modified) {
+							content[index] = val.default
+						} else if (!instance.originalStart && instance.start > datastoreVal.modified) {
+							content[index] = val.default
+						} else {
+							content[index] = datastoreVal.value
+						}
+					} else {
+						// todo - can we do a tighter check
+						fillState(val, instance)
+					}
+				}
+			})
+		}
+		Object.values(filledState.layers).forEach((layer) => {
+			fillState(layer.content, layer.instance)
+		})
+
+		return this.getDevice(deviceId)?.device.handleState(filledState, mappings)
+	}
+	setDatastore(newStore: Record<string, any>) {
+		this._actionQueue
+			.add(() => {
+				const allKeys = new Set([...Object.keys(newStore), ...Object.keys(this._datastore)])
+
+				const changed: string[] = []
+				for (const key of allKeys) {
+					if (this._datastore[key]?.value !== newStore[key]?.value) {
+						// it changed! let's sift through our dependencies to see if we need to do anything
+						Object.entries(this._deviceStates).forEach(([deviceId, states]) => {
+							if (states.find((state) => state.dependencies.find((deps) => deps === key))) {
+								changed.push(deviceId)
+							}
+						})
+					}
+				}
+
+				this._datastore = newStore
+
+				for (const deviceId of changed) {
+					const toBeFilled = _.compact([
+						this._deviceStates[deviceId].reverse().find((s) => s.time <= this.getCurrentTime()), // one state before now
+						...this._deviceStates[deviceId].filter((s) => s.time > this.getCurrentTime()), // all states after now
+					])
+
+					for (const s of toBeFilled) {
+						const filledState: typeof s.state = JSON.parse(JSON.stringify(s.state))
+						Object.values(filledState.layers).forEach(({ content, instance }) => {
+							if ((content as TSRTimelineObjBase['content']).$references) {
+								Object.entries((content as TSRTimelineObjBase['content']).$references || {}).forEach(([key, ref]) => {
+									const datastoreVal = this._datastore[key]
+									const set = (obj: Record<string, any>, path: string, val: any) => {
+										const p = path.split('.')
+										p.slice(0, -1).reduce((a, b) => a[b], obj)[p.slice(-1)[0]] = val
```

Comment:
Should this be in a try-catch? It seems like something that can crash the entire conductor.

Reply:
Yeah, could have definitely crashed it. I've updated the `set` method now so that it can't access undefined properties anymore. I think this is more useful anyway because it means we do not require a default for every value